salt/modules/napalm_formula.py

```diff
@@ -0,0 +1,367 @@
+# -*- coding: utf-8 -*-
+'''
+NAPALM Formula helpers
+======================
+
+.. versionadded:: Fluorine
+
+This is an Execution Module providing helpers for various NAPALM formulas,
+e.g., napalm-interfaces-formula, napalm-bgp-formula, napalm-ntp-formula etc.,
+meant to provide various helper functions to make the templates more readable.
+'''
+from __future__ import absolute_import, unicode_literals, print_function
+
+# Import python libs
+import copy
+import logging
+import fnmatch
+
+# Import salt modules
+import salt.utils.napalm
+import salt.ext.six as six
+import salt.utils.dictupdate
+from salt.defaults import DEFAULT_TARGET_DELIM
+try:
+    from salt.utils import traverse_dict_and_list
+except ImportError:
+    from salt.utils.data import traverse_dict_and_list
```

Comment:
Since this code is going into develop, there is no need to do this try/except, since the function lives in `salt.utils.data` now. Not only that, there are three problems with this approach:

1. Importing a function into the module's global scope will add it to the `__salt__` dunder (i.e. `__salt__['napalm_formula.traverse_dict_and_list']`)
2. This will also add this function to the docs, which will be confusing.
3. Since we have compatibility functions in `salt.utils` for all the funcs that were moved for 2018.3, the first import will succeed and this will result in deprecation warnings.

Instead, you should simply `import salt.utils.data` and invoke the function with its full name.

Reply:
When possible, it would be great to also use `__utils__['data.traverse_dict_list']()` because then it will allow people to include that data.py file in their `salt://_utils/` directory if they are unable to upgrade to the newer version of salt.

Reply:
Doing it @gtmanfred's way will also prevent you from needing to import the module at all.

Reply:
Thanks guys for reviewing this and your thoughts. As Daniel mentioned, I made these imports having backwards potability in mind; using `__utils__` sounds like a better idea, thought it would require backporting loads of util modules.

Re 2: Indeed, I thought about that too, but it seems like Sphinx is smart enough not to pick these functions (no idea why) and they don't appear in the documentation (tested with both Salt and out internal documentation).

Anyway, I will change to use `__utils__` as it seems a better tradeoff.

Reply:
Updated, though I just wanted to point this out: porting `utils.data` means it also requires porting `utils.dictupdate`, `utils.stringutils`, and `utils.yaml` (as they are imported there).

This is fine, though when I need, e.g., `files.fopen`, or `args.clean_kwargs`, I would also need to port loads of others such as `utils.jid`, `utils.path`, `utils.platform`, `utils.data`... and here's the catch: `salt.utils.args` needs `salt.utils.data` (which obviously won't be available). So not only that one basically needs to port almost all the modules under `utils/` but also they need to be patched (i.e., replace imports like the one from `utils.args` from `import salt.utils.data` to `import _utils.data`. This sounds like a nightmare to me.

Looking at this vs. 1 & 3 I'm not sure which is the worst...

Reply:
What about the following:

```python
from salt.version import __version_info__
if __version_info__ < (2018, 3):
    from salt.utils import traverse_list_and_dict as _traverse_list_and_dict
else:
    from salt.utils.data import traverse_list_and_dict as _traverse_list_and_dict
```

Or why not, even simpler:

```python
try:
    from salt.utils.data import traverse_dict_and_list as _traverse_dict_and_list
except ImportError:
    from salt.utils import traverse_dict_and_list as _traverse_dict_and_list
```

This way we can ensure that the function won't be exported (note that I flipped the import error to prevent deprecation warning thrown on >= Oxygen).

Does this sound like a better alternative?

Reply:
This does seem like a better alternative.

```python
try:
    from salt.utils.data import traverse_dict_and_list as _traverse_dict_and_list
except ImportError:
    from salt.utils import traverse_dict_and_list as _traverse_dict_and_list
```

However, I'm not sure why this should be necessary at all in `develop`.

Reply:
Yes, the question just comes down to, "How friendly to we want to be for allowing people to backport these modules?"

Reply:
Added https://github.com/saltstack/salt/pull/48879/commits/3e9ccdb397ba4fea30c97dba2084b91e44db7287 - please let me know if this is okay for the time being.